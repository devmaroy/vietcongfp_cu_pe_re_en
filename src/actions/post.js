// Post Actions

import database from '../firebase/firebase';
import history from '../routers/history';
import { 
    GET_POSTS,
    GET_POST,
    ADD_POST,
    EDIT_POST,
    REMOVE_POST,
    GET_COMMENTS,
    ADD_COMMENT,
    REMOVE_COMMENT,
    POST_LOADING,
    CLEAR_POST_LOADING,
    COMMENT_LOADING,
    CLEAR_COMMENT_LOADING,
    GET_ERRORS
} from './types';


// Get Posts
export const startGetPosts = () => {
    return ( dispatch ) => {
        dispatch( setPostLoading() );

        return database
            .ref( 'posts' )
            .orderByChild( 'createdAt' )
            .once( 'value' )
            .then( ( snapshot ) => {
                const posts = [];

                snapshot.forEach( ( childSnapshot ) => {
                    posts.push({
                        id: childSnapshot.key,
                        ...childSnapshot.val()
                    });
                });

                dispatch({
                    type: GET_POSTS,
                    posts
                });
    
                dispatch( clearPostLoading() );
            })
            .catch( ( err ) => {
                // Simple redirect to non existent route
                history.push( '/oops' );
            });
    };
};


// Get Post
export const startGetPost = ( id ) => {
    return ( dispatch ) => {
        dispatch( setPostLoading() );

        return database
            .ref( `posts/${ id }` )
            .once( 'value' )
            .then( ( snapshot ) => {
                if ( snapshot.exists() ) {
                    const post = {
                        id: snapshot.key,
                        ...snapshot.val()
                    };
        
                    dispatch({
                        type: GET_POST,
                        post
                    });
    
                    dispatch( clearPostLoading() );
                } else {
                    // Simple redirect to non existent route
                    history.push( '/not-found' );
                }
            })
            .catch( ( err ) => {
                // Simple redirect to non existent route
                history.push( '/oops' );
            });
    };
};


// Add Post
export const startAddPost = ( postData ) => {
    return ( dispatch ) => {
        return database
            .ref( 'posts' )
            .push( postData )
            .then( ( ref ) => {
                dispatch({
                    type: ADD_POST,
                    post: {
                        id: ref.key,
                        ...postData 
                    }
                });
    
                history.push( '/' );
            })
            .catch( ( err ) => {
                // Simple redirect to non existent route
                history.push( '/oops' );
            });
    };
} ;


// Edit Post 
export const startEditPost = ( id, updates ) => {
    return ( dispatch ) => {
        return database
            .ref( `posts/${ id }` )
            .update( updates )
            .then( ( ref ) => {
                dispatch({
                    type: EDIT_POST,
                    id,
                    updates
                });
    
                history.push( `/posts/${ id }` );
            })
            .catch( ( err ) => {
                // Simple redirect to non existent route
                history.push( '/oops' );
            });
    };
};


// Remove Post
export const startRemovePost = ( id ) => {
    return ( dispatch ) => {
        return database 
            .ref( `posts/${ id }` )
            .remove()
            .then( () => {
                dispatch({
                    type: REMOVE_POST,
                    id
                });
    
                history.push( '/' );
            })
            .catch( ( err ) => {
                // Simple redirect to non existent route
                history.push( '/oops' );
            });
    };
};


// Get Comments By Post Id
export const startGetCommentsByPostId = ( postId ) => {
    return ( dispatch ) => {
        dispatch( setCommentLoading() );

        return database
            .ref( 'comments' )
            .orderByChild( 'createdAt' )
            .once( 'value' )
            .then( ( snapshot ) => {
                const comments = [];
            
                snapshot.forEach( ( childSnapshot ) => {
                    if ( childSnapshot.val().postId === postId ) {
                        comments.push({
                            id: childSnapshot.key,
                            ...childSnapshot.val()
                        });
                    }
                });
    
                dispatch({
                    type: GET_COMMENTS,
                    comments
                });
    
                dispatch( clearCommentLoading() );
            })
            .catch( ( err ) => {
                // Simple redirect to non existent route
                history.push( '/oops' );
            });
    };
};


// Add Comment
export const startAddComment = ( commentData ) => {
    return ( dispatch ) => {
        return database
            .ref( 'comments' ) 
            .push( commentData )
            .then( ( ref ) => {
                dispatch({
                    type: ADD_COMMENT,
                    comment: {
                        id: ref.key,
                        ...commentData
                    }
                });
            })
            .catch( ( err ) => {
                // Simple redirect to non existent route
                history.push( '/oops' );
            });
    };
};


// Remove Comment
export const startRemoveComment = ( id ) => {
    return ( dispatch ) => {
        return database 
            .ref( `comments/${ id }` )
            .remove()
            .then(() => {
                dispatch({
                    type: REMOVE_COMMENT,
                    id
                });
            })
            .catch( ( err ) => {
                // Simple redirect to non existent route
                history.push( '/oops' );
            });
    };
};


// Post Loading
export const setPostLoading = () => ({
    type: POST_LOADING
});


// Clear Post Loading
export const clearPostLoading = () => ({
    type: CLEAR_POST_LOADING
});


// Comment Loading
export const setCommentLoading = () => ({
    type: COMMENT_LOADING
});


// Clear Comment Loading
export const clearCommentLoading = () => ({
    type: CLEAR_COMMENT_LOADING
});