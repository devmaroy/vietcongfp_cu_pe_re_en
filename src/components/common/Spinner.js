import React from 'react';
import PropTypes from 'prop-types';

const Spinner = ( { imagePath, alt, style, classNameWrapper, classNameImg } ) => (
    <div className={ classNameWrapper }>
        <img 
            style={ style }
            src={ imagePath } 
            alt={ alt }
            className={ classNameImg }
        />
    </div>
);

Spinner.propTypes = {
    imagePath: PropTypes.string,
    alt: PropTypes.string,
    style: PropTypes.object,
    classNameWrapper: PropTypes.string,
    classNameImg: PropTypes.string
};

Spinner.defaultProps = {
    imagePath: '/images/loader.gif',
    alt: 'Loading...',
    style: { width: '80px', margin: '60px', display: 'block' }
};

export default Spinner;