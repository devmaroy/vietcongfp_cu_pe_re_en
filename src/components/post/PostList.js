import React from 'react';
import PropTypes from 'prop-types';
import Pagination from "react-js-pagination";
import PostItem from './PostItem';

class PostList extends React.Component {
    state = {
        activePage: 1,
        postsPerPage: 5
    }

    handlePageChange = ( pageNumber ) => {
        this.setState( () => ( { activePage: pageNumber } ) );
    }

    render() {
        const { activePage, postsPerPage } = this.state;
        const { posts } = this.props;

        return (
            <div>
                {
                    posts.length === 0 ? (
                        <p>There are no posts.</p>
                    ) : (
                        <div>
                            <div className="posts">
                                {
                                    posts
                                        .slice( ( activePage - 1 ) * postsPerPage, ( postsPerPage * activePage ) )
                                        .map( ( post ) => <PostItem key={ post.id } post={ post } /> )
                                }
                            </div>
                            {
                                posts.length > postsPerPage && (
                                    <div className="site-pagination">
                                        <Pagination
                                            activePage={ this.state.activePage }
                                            itemsCountPerPage={ postsPerPage }
                                            totalItemsCount={ posts.length }
                                            pageRangeDisplayed={ 5 }
                                            onChange={ this.handlePageChange }
                                        />
                                    </div>
                                )
                            }
                        </div>
                    )
                }
            </div>
        );
    };
};

PostList.propTypes = {
    posts: PropTypes.array.isRequired
};

export default PostList;