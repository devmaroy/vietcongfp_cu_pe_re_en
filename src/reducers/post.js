// Post Reducer

import {
    GET_POSTS,
    GET_POST,
    ADD_POST,
    EDIT_POST,
    REMOVE_POST,
    GET_COMMENTS,
    ADD_COMMENT,
    REMOVE_COMMENT,
    SET_POST_LOADING,
    CLEAR_POST_LOADING,
    SET_COMMENT_LOADING,
    CLEAR_COMMENT_LOADING
} from '../actions/types';

const postReducerDefaultState = {
    posts: [],
    post: {},
    comments: [],
    isPostLoading: false,
    isCommentLoading: false
};

export default ( state = postReducerDefaultState, action ) => {
    switch ( action.type ) {
        
        case GET_POSTS:
            return {
                ...state,
                posts: action.posts.reverse()
            };

        case GET_POST:
            return {
                ...state,
                post: action.post
            };

        case ADD_POST:
            return {
                ...state,
                posts: [ action.post, ...state.posts ]
            };

        case EDIT_POST:
            return {
                ...state,
                posts: state.posts.map( ( post ) => {
                    if ( post.id === action.id ) {
                        return {
                            ...post,
                            ...action.updates
                        }
                    } else {
                        return post;
                    }
                })
            };

        case REMOVE_POST:
            return {
                ...state,
                posts: state.posts.filter( ( post ) => post.id !== action.id )
            };
        
        case GET_COMMENTS: 
            return {
                ...state,
                comments: action.comments.reverse()
            }
            
        case ADD_COMMENT:
            return {
                ...state,
                comments: [ action.comment, ...state.comments ]
            };

        case REMOVE_COMMENT:
            return {
                ...state,
                comments: state.comments.filter( ( comment ) => comment.id !== action.id )
            };

        case SET_POST_LOADING:
            return {
                ...state,
                isPostLoading: true
            };

        case SET_COMMENT_LOADING:
            return {
                ...state,
                isCommentLoading: true
            };

        case CLEAR_COMMENT_LOADING:
            return {
                ...state,
                isCommentLoading: false
            };

        case CLEAR_POST_LOADING:
            return {
                ...state,
                isPostLoading: false
            };
    
        default:
            return state;
    };
};