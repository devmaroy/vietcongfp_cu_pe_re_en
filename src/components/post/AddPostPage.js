import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { startAddPost } from '../../actions/post';
import PostForm from './PostForm';

export class AddPostPage extends React.Component {
    onSubmit = ( postData ) => {
        this.props.startAddPost( postData );
    };

    render() {
        return (
            <div className="content-container">
                <div className="page">
                    <h2>Create Post</h2>
                    <PostForm 
                        onSubmit={ this.onSubmit }
                    />
                </div>
            </div>
        )
    };
};

AddPostPage.propTypes = {
    startAddPost: PropTypes.func.isRequired
};

const mapDispatchToProps = ( dispatch ) => ({
    startAddPost: ( postData ) => dispatch( startAddPost( postData ) )
});

export default connect( undefined, mapDispatchToProps )( AddPostPage );