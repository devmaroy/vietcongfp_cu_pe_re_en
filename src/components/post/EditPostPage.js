import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { startGetPost, startEditPost, startRemovePost } from '../../actions/post';
import Spinner from '../common/Spinner';
import PostForm from './PostForm';

class EditPostPage extends React.Component {
    componentDidMount() {
        this.props.startGetPost( this.props.match.params.id );
    };

    onRemove = () => {
        this.props.startRemovePost( this.props.match.params.id );
    };

    onSubmit = ( updates ) => {
        this.props.startEditPost( this.props.match.params.id, updates );
    };

    render() {
        const { post, isPostLoading } = this.props.post;
        let postEditContent;

        if ( post === null || isPostLoading || Object.keys( post ).length === 0 ) {
            postEditContent = <Spinner />
        } else {
           postEditContent = (
              <div>
                    <h3>Edit Post</h3>
                    <PostForm
                        onSubmit={ this.onSubmit }
                        post={ post }
                    />
                    <button onClick={ this.onRemove }>Remove</button>
              </div>
           );
        }

        return (
            <div>
                { postEditContent }
            </div>
        );
    }
}

EditPostPage.propTypes = {
    startGetPost: PropTypes.func.isRequired,
    startRemovePost: PropTypes.func.isRequired,
    startEditPost: PropTypes.func.isRequired,
    post: PropTypes.object.isRequired
};

const mapStateToProps = ( state ) => ({
    post: state.post
});

const mapDispatchToProps = ( dispatch ) => ({
    startGetPost: ( id ) => dispatch( startGetPost( id ) ),
    startEditPost: ( id, updates ) => dispatch( startEditPost( id, updates ) ),
    startRemovePost: ( id ) => dispatch( startRemovePost( id ) )
});

export default connect( mapStateToProps, mapDispatchToProps )( EditPostPage );