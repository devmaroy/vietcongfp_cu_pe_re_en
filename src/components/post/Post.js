import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { startGetPostById } from '../../actions/post';
import Spinner from '../common/Spinner';
import PostItem from './PostItem';
import CommentPage from '../comment/CommentPage';

export class Post extends React.Component {
    componentDidMount() {
        this.props.startGetPostById( this.props.match.params.id );
    };

    render() {
        const { post, isPostLoading } = this.props.post;
        let postContent;

        if ( post === null || isPostLoading || Object.keys( post ).length === 0 ) {
            postContent = <Spinner />;       
        } else {
            postContent = (
                <div>
                    <PostItem key={ post.id } post={ post } showActions={ false } isSingle={ true } />
                    <CommentPage postId={ post.id } />
                </div>
            )
        }

        return (
            <div className="content-container">
                <div className="page">
                    { postContent }
                </div>
            </div>
        );
    }
}

Post.propTypes = {
    startGetPostById: PropTypes.func.isRequired,
    post: PropTypes.object.isRequired
};

const mapStateToProps = ( state ) => ({
    post: state.post
});

const mapDispatchToProps = ( dispatch ) => ({
    startGetPostById: ( id ) => dispatch( startGetPostById( id ) )
});

export default connect( mapStateToProps, mapDispatchToProps )( Post );     