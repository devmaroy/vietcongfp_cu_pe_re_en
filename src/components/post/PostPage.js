import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { startGetPosts } from '../../actions/post';
import Spinner from '../common/Spinner';
import PostList from './PostList';

export class PostPage extends React.Component {
    componentDidMount() {
        this.props.startGetPosts();
    };

    render() {
        const { posts, isPostLoading } = this.props.post;
        let postsContent;

        if ( posts === null || isPostLoading ) {
            postsContent = <Spinner />;
        } else {
           postsContent = <PostList posts={ posts } />;
        }

        return (
            <div className="content-container">
                <div className="page">
                    { postsContent }
                </div>
            </div>
        );
    }
}

PostPage.propTypes = {
    startGetPosts: PropTypes.func.isRequired,
    post: PropTypes.object.isRequired
};

const mapStateToProps = ( state ) => ({
    post: state.post
});

const mapDispatchToProps = ( dispatch ) => ({
    startGetPosts: () => dispatch( startGetPosts() )
});

export default connect( mapStateToProps, mapDispatchToProps )( PostPage );