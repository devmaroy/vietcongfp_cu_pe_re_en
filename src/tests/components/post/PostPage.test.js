import React from 'react';
import { shallow } from 'enzyme';
import { PostPage } from '../../../components/post/PostPage';
import Spinner from '../../../components/common/Spinner';
import PostList from '../../../components/post/PostList';

test( 'should render correctly PostPage with spinner', () => {
    const post = {
        posts: [],
        isPostLoading: true
    };

    const wrapper = shallow(
        <PostPage
            post={ post }
            startGetPosts={ () => {} }
        />
    );

    expect( wrapper ).toMatchSnapshot();
    expect( wrapper.find( Spinner ).exists() ).toBeTruthy();
});

test( 'should not render PostPage with spinner', () => {
    const post = {
        posts: [],
        isPostLoading: false
    };

    const wrapper = shallow(
        <PostPage
            post={ post }
            startGetPosts={ () => {} }
        />
    );

    expect( wrapper.find( Spinner ).exists() ).toBeFalsy();
})

test( 'should render correctly PostList', () => {
    const post = {
        posts: [],
        isPostLoading: false
    };

    const wrapper = shallow(
        <PostPage
            post={ post }
            startGetPosts={ () => {} }
        />
    );

    expect( wrapper ).toMatchSnapshot();
    expect( wrapper.find( PostList ).exists() ).toBeTruthy();
});