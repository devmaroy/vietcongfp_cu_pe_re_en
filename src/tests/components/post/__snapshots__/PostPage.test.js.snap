// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly PostList 1`] = `
<div
  className="content-container"
>
  <div
    className="page"
  >
    <PostList
      posts={Array []}
    />
  </div>
</div>
`;

exports[`should render correctly PostPage with spinner 1`] = `
<div
  className="content-container"
>
  <div
    className="page"
  >
    <Spinner
      alt="Loading..."
      imagePath="/images/loader.gif"
      style={
        Object {
          "display": "block",
          "margin": "60px",
          "width": "80px",
        }
      }
    />
  </div>
</div>
`;
